### PR TITLE
feat: Cookie same site should accept function, string, or None [APP_INFO #2]

### PR DESCRIPTION
**Pull request by iresharma**
## Summary of change

(A few sentences about this PR)

## Related issues

-   Link to issue1 here
-   Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py
-   [ ] Make sure that `syncio` / `asyncio` functions are consistent.
 
## Remaining TODOs for this PR

-   [ ] Item1
-   [ ] Item2

### CODE REVIEW

File: supertokens_python/recipe/session/cookie_and_header.py [hunk 1]
```diff
@@ -179,27 +181,30 @@ def clear_session_from_all_token_transfer_methods(
...
 
-def clear_session_mutator(config: SessionConfig, transfer_method: TokenTransferMethod):
+def clear_session_mutator(
+    request: BaseRequest, config: SessionConfig, transfer_method: TokenTransferMethod
+):
     def mutator(
         response: BaseResponse,
     ):
-        return _clear_session(response, config, transfer_method)
+        return _clear_session(request, response, config, transfer_method)
```
Comment (KShivendu) [inline supertokens_python/recipe/session/cookie_and_header.py:193]:
```suggestion
        return await _clear_session(request, response, config, transfer_method)
```

Comment (KShivendu) [inline supertokens_python/recipe/session/cookie_and_header.py:193]: This `_clear_session` async. So use async/await here.

Comment (KShivendu) [reply supertokens_python/recipe/session/cookie_and_header.py:193]: Please also check all such occurrences and fix similar mistakes.

File: supertokens_python/recipe/session/cookie_and_header.py [hunk 2]
```diff
@@ -211,13 +216,14 @@ def _clear_session(
...
 def clear_session_response_mutator(
+    request: BaseRequest,
     config: SessionConfig,
     transfer_method: TokenTransferMethod,
 ):
     def mutator(
         response: BaseResponse,
     ):
-        return _clear_session(response, config, transfer_method)
+        return _clear_session(request, response, config, transfer_method)
```
Comment (KShivendu) [inline supertokens_python/recipe/session/cookie_and_header.py:226]: Use async def mutator and await _clear_session(...)

File: supertokens_python/recipe/session/cookie_and_header.py [hunk 3]
```diff
@@ -142,7 +144,7 @@ def set_cookie_response_mutator(
     def mutator(
         response: BaseResponse,
     ):
-        return _set_cookie(response, config, key, value, expires, path_type)
+        return _set_cookie(request, response, config, key, value, expires, path_type)
```
Comment (KShivendu) [inline supertokens_python/recipe/session/cookie_and_header.py:147]: `_set_cookie` is an async function. Use async/await here.

File: supertokens_python/recipe/session/interfaces.py [hunk 4]
```diff
@@ -362,7 +362,7 @@ async def verify_session(
         pass
 
 
-ResponseMutator = Callable[[BaseResponse], None]
+ResponseMutator = Callable[[BaseResponse], MaybeAwaitable[None]]
```
Comment (KShivendu) [inline supertokens_python/recipe/session/interfaces.py:365]: Maybe just use async functions everywhere? I don't think there's any harm making all mutators async. But there's a big plus: linter will prbaobly catch any mistakes for now and for future (verify this once)

```suggestion
ResponseMutator = Callable[[BaseResponse], Awaitable[None]]
```

File: supertokens_python/recipe/session/session_class.py [hunk 5]
```diff
@@ -125,9 +124,6 @@ async def update_session_data_in_database(
             self.session_handle, new_session_data, user_context
         )
         if not updated:
-            log_debug_message(
-                "updateSessionDataInDatabase: Throwing UNAUTHORISED because session does not exist anymore"
-            )
```
Comment (KShivendu) [inline supertokens_python/recipe/session/session_class.py:130]: why removing these logs?

File: supertokens_python/recipe/session/utils.py [hunk 6]
```diff
@@ -391,21 +401,42 @@ def validate_and_normalise_user_input(
...
+            top_level_api_domain != top_level_origin
+            or api_domain_scheme != origin_scheme
+        ):
+            cookie_same_site_ret = "none"
+        else:
+            cookie_same_site_ret = "lax"
+
+        if cookie_same_site is not None:
+            return cookie_same_site_normalize
+        return cookie_same_site_ret
```
Comment (KShivendu) [inline supertokens_python/recipe/session/utils.py:439]:
```suggestion
        return result
```

*_ret* is kinda bad for being part of var name.

**Review by KShivendu — CHANGES_REQUESTED** [review]
Consider thinking if we can make `ResponseMutator = Callable[[BaseResponse], Awaitable[None]]` since it gives the advantage that linter will catch mistakes (i.e. async function called without await) that it wont otherwise. (confirm this please)

File: supertokens_python/framework/django/django_middleware.py [hunk 7]
```diff
@@ -45,7 +45,7 @@ async def __asyncMiddleware(request: HttpRequest):
                 if hasattr(request, "supertokens") and isinstance(
                     request.supertokens, SessionContainer  # type: ignore
                 ):
-                    manage_session_post_response(
+                    await manage_session_post_response(
```
Comment (nkshah2) [inline supertokens_python/framework/django/django_middleware.py:48]: Wait does this mean this was a bug before?

File: supertokens_python/framework/fastapi/fastapi_middleware.py [hunk 8]
```diff
@@ -56,7 +56,9 @@ async def dispatch(self, request: Request, call_next: RequestResponseEndpoint):
                 if hasattr(request.state, "supertokens") and isinstance(
                     request.state.supertokens, SessionContainer
                 ):
-                    manage_session_post_response(request.state.supertokens, result)
+                    await manage_session_post_response(
```
Comment (nkshah2) [inline supertokens_python/framework/fastapi/fastapi_middleware.py:59]: Does this not need a async to sync version for fastapi?

File: supertokens_python/framework/flask/flask_middleware.py [hunk 9]
```diff
@@ -65,7 +65,7 @@ def _(response: Response):
 
             response_ = FlaskResponse(response)
             if hasattr(g, "supertokens") and g.supertokens is not None:
-                manage_session_post_response(g.supertokens, response_)
+                sync(manage_session_post_response(g.supertokens, response_))
```
Comment (nkshah2) [inline supertokens_python/framework/flask/flask_middleware.py:68]: Does flask not need an async version?

File: supertokens_python/recipe/session/recipe.py [hunk 10]
```diff
@@ -77,7 +78,11 @@ def __init__(
         app_info: AppInfo,
         cookie_domain: Union[str, None] = None,
         cookie_secure: Union[bool, None] = None,
-        cookie_same_site: Union[Literal["lax", "none", "strict"], None] = None,
+        cookie_same_site: Union[
+            Callable[[BaseRequest, Any], Literal["lax", "none", "strict"]],
```
Comment (nkshah2) [inline supertokens_python/recipe/session/recipe.py:82]: We should typedef this

Comment (nkshah2) [reply supertokens_python/recipe/session/recipe.py:82]: Just `"lax", "none", "strict"` I mean

File: supertokens_python/recipe/session/session_class.py [hunk 11]
```diff
@@ -36,7 +36,7 @@
 )
 from .constants import protected_props
 from ...framework import BaseRequest
-from supertokens_python.utils import log_debug_message
+from ...logger import log_debug_message
```
Comment (nkshah2) [inline supertokens_python/recipe/session/session_class.py:39]: Do you need this import?

Comment (nkshah2) [reply supertokens_python/recipe/session/session_class.py:39]: If you do change it back to the original import

File: supertokens_python/recipe/session/utils.py [hunk 12]
```diff
@@ -428,7 +459,7 @@ def validate_and_normalise_user_input(
         )
 
     if anti_csrf is None:
-        anti_csrf = "VIA_CUSTOM_HEADER" if cookie_same_site == "none" else "NONE"
+        anti_csrf = "VIA_CUSTOM_HEADER"
```
Comment (nkshah2) [inline supertokens_python/recipe/session/utils.py:462]: Merge the base branch into this one so that code changes from other PRs dont show here

**Review by nkshah2 — CHANGES_REQUESTED** [review]

File: supertokens_python/supertokens.py [hunk 13]
```diff
@@ -133,10 +134,12 @@ def defaultImpl(o: Any):
...
 
 
-def manage_session_post_response(session: SessionContainer, response: BaseResponse):
+async def manage_session_post_response(
+    session: SessionContainer, response: BaseResponse
+):
     # Something similar happens in handle_error of session/recipe.py
     for mutator in session.response_mutators:
-        mutator(response=response)  # type: ignore
+        await resolve(mutator(response=response))  # type: ignore
```
Comment (KShivendu) [inline supertokens_python/supertokens.py:142]: Now that all mutators must be async, you don't need `resolve`, right?

Comment (KShivendu) [inline supertokens_python/supertokens.py:142]: Now that all mutators must be async, you don't need resolve, right?

Comment (KShivendu) [reply supertokens_python/supertokens.py:142]: There's one more occurence of this. Fix that as well.

File: supertokens_python/recipe/session/session_class.py (hunk 5 above)
Comment (KShivendu) [reply supertokens_python/recipe/session/session_class.py]: @iresharma ?

File: supertokens_python/supertokens.py (hunk 13 above)
Comment (KShivendu) [reply supertokens_python/supertokens.py:142]: @iresharma ^

**Review by KShivendu — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/session/session_class.py (hunk 5 above)
Comment (iresharma) [reply supertokens_python/recipe/session/session_class.py]: It was accidental, I added it back

File: supertokens_python/supertokens.py (hunk 13 above)
Comment (iresharma) [reply supertokens_python/supertokens.py]: Done

File: supertokens_python/recipe/session/recipe.py [hunk 14]
```diff
@@ -77,7 +78,11 @@ def __init__(
         app_info: AppInfo,
         cookie_domain: Union[str, None] = None,
         cookie_secure: Union[bool, None] = None,
-        cookie_same_site: Union[Literal["lax", "none", "strict"], None] = None,
+        cookie_same_site: Union[
```
Comment (nkshah2) [inline supertokens_python/recipe/session/recipe.py:81]: Update the CHANGELOG for all user facing changes in this PR

**Review by nkshah2 — CHANGES_REQUESTED** [review]

**Pull request merged**